File: src/main/java/me/cephetir/skyskipped/utils/PingUtils.java
/*
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * Version 2, December 2004
 *
 * Copyright (C) 2021 Cephetir
 *
 * Everyone is permitted to copy and distribute verbatim or modified
 * copies of this license document, and changing it is allowed as long
 * as the name is changed.
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 */

package me.cephetir.skyskipped.utils;

import gg.essential.universal.ChatColor;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

public class PingUtils {
    private final Minecraft mc = Minecraft.getMinecraft();
    private int ticks;
    private final String text;

    public PingUtils(int ticks, String text) {
        this.ticks = ticks;
        this.text = text;
        MinecraftForge.EVENT_BUS.register(this);
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if (ticks <= 0) MinecraftForge.EVENT_BUS.unregister(this);
        ticks--;
    }

    @SubscribeEvent
    public void draw(RenderGameOverlayEvent.Text event) {
        if(ticks % 5 == 0) mc.thePlayer.playSound("random.orb", 1f, 1f);
        GlStateManager.pushMatrix();
        GlStateManager.scale(1.5f, 1.5f, 1.5f);
        mc.fontRendererObj.drawStringWithShadow(ChatColor.DARK_RED + text,
                event.resolution.getScaledWidth() / 1.5f / 2f - mc.fontRendererObj.getStringWidth(text) / 2f,
                event.resolution.getScaledHeight() / 1.5f / 2f - 6.75f, -1);
        GlStateManager.popMatrix();
    }
}

File: src/main/kotlin/me/cephetir/skyskipped/utils/skyblock/PingUtils.kt
/*
 * SkySkipped - Hypixel Skyblock QOL mod
 * Copyright (C) 2023  Cephetir
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.cephetir.skyskipped.utils.skyblock

import me.cephetir.bladecore.core.event.BladeEventBus
import me.cephetir.bladecore.core.event.listener.listener
import me.cephetir.bladecore.utils.player
import me.cephetir.bladecore.utils.threading.safeListener
import me.cephetir.bladecore.utils.world
import net.minecraft.client.renderer.GlStateManager
import net.minecraftforge.client.event.RenderGameOverlayEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent

class PingUtils(private var ticks: Int, private val text: String, private val sound: Boolean = true, private val dynamicText: (() -> String)? = null, private val callback: () -> Unit = {}) {

    init {
        listener<ClientTickEvent> {
            if (it.phase != TickEvent.Phase.START) return@listener
            if (world == null || player == null) return@listener BladeEventBus.unsubscribe(this)

            if (ticks <= 0) {
                callback()
                BladeEventBus.unsubscribe(this)
            }
            if (sound && ticks % 2 == 0) player!!.playSound("random.orb", 1f, 1f)
            ticks--
        }

        safeListener<RenderGameOverlayEvent.Text> {
            GlStateManager.pushMatrix()
            GlStateManager.scale(1.5f, 1.5f, 1.5f)
            val displayText = dynamicText?.invoke() ?: text
            mc.fontRendererObj.drawStringWithShadow(
                "§4§l$displayText",
                it.resolution.scaledWidth / 1.5f / 2f - mc.fontRendererObj.getStringWidth(displayText) / 2f,
                it.resolution.scaledHeight / 1.5f / 2f - 6.75f, -1
            )
            GlStateManager.scale(1f / 1.5f, 1f / 1.5f, 1f / 1.5f)
            GlStateManager.popMatrix()
        }

        BladeEventBus.subscribe(this)
    }
}